pipelines/batch/src/main/java/com/google/fhir/analytics/metrics/FlinkPipelineMetrics.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2020-2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.fhir.analytics.metrics;
+
+import com.google.fhir.analytics.MetricsConstants;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import org.apache.beam.runners.core.metrics.MetricsContainerStepMap;
+import org.apache.beam.runners.flink.metrics.FlinkMetricContainer;
+import org.apache.beam.sdk.metrics.MetricNameFilter;
+import org.apache.beam.sdk.metrics.MetricQueryResults;
+import org.apache.beam.sdk.metrics.MetricResult;
+import org.apache.beam.sdk.metrics.MetricResults;
+import org.apache.beam.sdk.metrics.MetricsFilter;
+import org.apache.flink.core.execution.JobClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FlinkPipelineMetrics implements PipelineMetrics {
+
+  private static final Logger logger = LoggerFactory.getLogger(FlinkPipelineMetrics.class);
+
+  private static ConcurrentHashMap<String, JobClient> jobClientMap = new ConcurrentHashMap<>();
```

Comment:
I have not looked at all the details of Metrics related changes yet but is it fair to say that the new logic is basically very similar to the old one but instead of assuming that we only have one pipeline (and hence one "active" `jobClient`) we may have multiple now because there might be multiple pipelines?

Also the `static` requirements here are because of how `FlinkJobListener` is implemented correct? I wonder if it is possible to somehow set a new instance of  `FlinkPipelineMetrics` inside job-listener whenever we run a new batch of pipelines (basically I am trying to see if we can avoid this singleton pattern).

Reply:
Yes, your understanding correct. The logic is similar to the older one, just that now we rely on the metrics of all pipelines submitted as per the current batch.

As per the implementation of the `FlinkJobListener` a new instance of it is created for every new pipeline, since for every batch we create multiple pipelines, setting one instance of `FlinkPipelineMetrics` for every instance of `FlinkJobListener` will not work.

Reply:
Yes, having a static instance in the `PipelineMetricsProvider` provides the same desired effect. Hence the changes have been made accordingly. Also, the comments have been updated in the `FlinkPipelineMetrics` class.